python/ray/dag/compiled_dag_node.py

```diff
@@ -746,6 +743,15 @@ def _get_node_id(self):
             for idx in task.downstream_node_idxs:
                 frontier.append(idx)
 
+        from ray.dag.constants import RAY_ADAG_ENABLE_VERIFY_GRAPH
+
+        if RAY_ADAG_ENABLE_VERIFY_GRAPH:
+            assert self._verify_graph(), (
+                "Detect a deadlock in the graph. If this is a false positive, "
```

Comment:
Can you change this to a ValueError instead of assertion error (assertion error is normally used for unexpected conditions that indicate possibility of a system bug)? Also, can you improve this error message? Ideally we should list which actor tasks would cause the deadlock, and the message should ideally be a bit more clear, like:
"""
This DAG cannot be compiled because it will deadlock on NCCL calls. A.foo requires a result from B.bar but B.foo requires a result from A.bar ...
"""

Reply:
I added a log for all indices of DAG nodes that are not in the topological order. However, another question is that users don't know what the indices map to in the DAG nodes. Maybe we should provide a visualization tool in the future so that users can know how does a DAG looks like and the index of a DAG node.

Reply:
Updated https://github.com/ray-project/ray/pull/45960/commits/4babe2294f4a95cdb8e2f888727464cab58e72e9